### DISCUSSION TITLE
In async server is not possible to disconnect client while sending him error message?

### DISCUSSION THREAD

**Discussion by MaksymSofer** [miguelgrinberg/python-socketio#805] [Q&A]
I`m using asyncServer implementation.
I want to validate JWT token from the client and if it is not valid, disconnect him with the error message being sent to him.
Is it not possible?

In my current implementation, i`m throwing 
raise ConnectionRefusedError('Close')

I've assumed that this should send "Close" message and disconnect the connection.
But instead, it just sending to the client error message " Error:Close" 

Version being used 5.4.0
Python 3.8

**Comment by miguelgrinberg** [comment]
Yes, the Error packet is the mechanism to deliver errors to the client, which is then supposed to end the connection.

I'm not sure there is a proper way to do it the way you suggest, but I'm open to improving this. The tricky part is to find a way to do this for HTTP and WebSocket, and make sure other Socket.IO implementations work well and are able to expose the error to the client application.

Also, please don't write bugs against this package when you need to discuss. This is what "Discussions" is for.

**Reply by MaksymSofer** [reply]
Sorry for that, didn't payed enough attention.

In WSGI - I assume that 
sio.emit(text)
sio.disconnect(sid)

Will do exactly that I'm trying to achieve, right?

**Reply by miguelgrinberg** [reply]
It's tricky. The emit needs to arrive at the client before you can disconnect. A lot of the work to emit happens in the background after the `emit()` method returns. You can use a callback to receive an acknowledgement for the emit, and then you can disconnect.

**Reply by MaksymSofer** [reply]
Per your suggestion i built the following:

async def method():
           await sio.emit("connection_closed", to=sid, callback=disconnect(sid))

async def disconnect(sid):
    await sio.disconnect(sid)
    
    
 But connection is still alive and in this case i`m getting "connection_closed" message
 I do something wrong?

**Reply by MaksymSofer** [reply]
moreover, then I stop Uvicorn  the following warning appear:
venv/lib/python3.8/site-packages/socketio/base_manager.py:108: RuntimeWarning: coroutine 'disconnect' was never awaited
  del self.callbacks[sid]
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

Loos like the disconnect is not being executed at all..

**Reply by miguelgrinberg** [reply]
The `callback` argument takes a function, not a coroutine or task.

```
async def method(sid):
    async def disconnect():
        await sio.disconnect(sid)

    await sio.emit("connection_closed", to=sid, callback=disconnect)
```

**Reply by MaksymSofer** [reply]
The client getting: connection_closed [null]

I modified code a bit to see if disconnect method is executed at all:

            await sio.emit("connection_closed", to=sid, callback=disconnect)


async def disconnect(sid):
    print("disconnect reached")
    await sio.disconnect(sid)
    
  It`s not - the print in it not being executed at all

**Reply by miguelgrinberg** [reply]
How does this work? Who provides the `sid` argument to the callback function?

**Reply by MaksymSofer** [reply]
it does not matter - function not being called at all.
I tried with disconnect(sid) too - the same result

**Reply by miguelgrinberg** [reply]
Make a complete runnable example that demonstrates the problem and I'll have a look at it. Every single example of code that you showed me so far has errors. I can't really answer your question when you don't show me actual working code.

**Comment by MaksymSofer** [comment]
```
async def disconnect(sid):
    print("disconnect reached")
    await sio.disconnect(sid)

    
@sio.event
async def connect(sid, environ, auth):
    await sio.emit("connection_closed", to=sid, callback=disconnect(sid))

```

This is stripped version of running and working code
as you can see, i added print into disconnect function and it is not executed.
If i change disconnect function from async to normal (and await removed from sio.disconnect ) - than the above text is printed - but still the connection is not closed

**Reply by miguelgrinberg** [reply]
We are going back in circles. The callback argument takes a function as an argument, not a coroutine. Please send me a complete example that I can run here, with a client and a server. Do not send snippets please. I also showed you how to create a callback that is async earlier and you have ignored it.

**Comment by MaksymSofer** [comment]
I found a way to safely close the connection:
I built as you suggested normal function (not a coroutine or a task) which add SID to array
Then after sio.emit sends error message to the corresponding client, it than doing "for" on the above array
disconnecting each client from array and removing SID`s

thank you for assistance
